mods/prevent-focus-stealing.wh.cpp

```diff
@@ -0,0 +1,371 @@
+
+*/
+// ==/WindhawkModReadme==
+
+// ==WindhawkModSettings==
+/*
+- applications:
+  - - exes: [electron.exe, brave.exe, chrome.exe]
+      $name: Application Executables
+      $description: "Executables cannot be blank. Each executable can be either a file name (e.g. electron.exe), full path (e.g. C:\\Program Files\\Brave\\brave.exe), or path glob (e.g. *electron.exe)."
+    - args: ['']
+      $name: Required Arguments
+      $description: "Each argument is a substring to match in the command line. All must match. Leave empty to include all processes."
+    - exclude_args: [--type]
+      $name: Excluded Arguments
+      $description: "Each argument is a substring to exclude from the command line. Any match excludes the process. Leave empty to include all processes."
+    - title: Developer Tools
+      $name: Window Title
+      $description: "The window title prefix to match. Leave blank to match all windows for the process."
+    - never_focus: false
+      $name: Never Focus
+      $description: "Never allow the window to be become focused, instead of only when it initially opens."
+  $name: Application Windows
+  $description: "Prevent focus being stolen by these application windows."
+*/
+// ==/WindhawkModSettings==
+
+#include <windhawk_utils.h>
+#include <windows.h>
+#include <shlwapi.h>
+#include <vector>
+#include <set>
+#include <map>
+
+#define WM_USER_REMOVE_NOACTIVATE (WM_USER + 1)
+
+const uint64_t NEW_WINDOW_TRACKING_MS = 1000;
+const uint64_t INITIAL_UI_FOCUS_BLOCK_MS = 100;
+const uint64_t INITIAL_FRAME_ACTIVATE_BLOCK_MS = 8000;
+
+struct WindowInfo {
+    const wchar_t* title;
+    bool neverFocus;
+};
+
+std::vector<WindowInfo> g_includedWindows;
+
+std::map<HWND, uint64_t> g_newHwnds;
+std::set<HWND> g_hookedHwnds;
+std::map<HWND, uint64_t> g_blockedHwnds; // Store creation time for timeout
+
+WindowInfo* GetWindowInfo(HWND hWnd, LPCWSTR title) {
+    for (WindowInfo& info : g_includedWindows) {
+        if (*info.title && _wcsnicmp(title, info.title, wcslen(info.title)) != 0)
+            continue;
+        if (!info.neverFocus && g_hookedHwnds.contains(hWnd))
+            continue;
+        return &info;
+    }
+    return nullptr;
+}
+
+LRESULT CALLBACK WindowSubclassProc(HWND hWnd, UINT uMsg, WPARAM wParam, LPARAM lParam, DWORD_PTR dwRefData) {
+    if (uMsg == WM_DESTROY) {
+        Wh_Log(L"Stopped tracking window 0x%p due to WM_DESTROY", hWnd);
+        g_newHwnds.erase(hWnd);
+        g_hookedHwnds.erase(hWnd);
+        g_blockedHwnds.erase(hWnd);
+    }
+    if (uMsg == WM_USER_REMOVE_NOACTIVATE) {
+        Wh_Log(L"Removing WS_EX_NOACTIVATE from window 0x%p", hWnd);
+        SetWindowLongPtrW(hWnd, GWL_EXSTYLE, GetWindowLongPtrW(hWnd, GWL_EXSTYLE) & ~WS_EX_NOACTIVATE);
+    }
+    if (uMsg == WM_ACTIVATE && (wParam == WA_ACTIVE || wParam == WA_CLICKACTIVE)) {
+        auto it = g_blockedHwnds.find(hWnd);
+        if (it != g_blockedHwnds.end() && GetTickCount64() - it->second < INITIAL_UI_FOCUS_BLOCK_MS) {
```

Comment:
I'm not sure about the chances for real issues, but `g_blockedHwnds` may be accessed from different thread, which is unsafe. Since it's used to store window-specific data, you can use `SetProp`/`GetProp`, which should also result in simpler code.

Reply:
Any thread can call `SetWindowTextW`, and several threads can have a message loop. The simplest example is: create two threads, and call MessageBox in each,

BTW `SetWindowTextW` can be called across processes too, and there are other way to set a window text (e.g. `WM_SETTEXT`).